# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [238]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [239]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [240]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [241]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [242]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [243]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [244]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [245]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [246]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [247]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [248]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [249]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [250]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [251]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [252]:
data.duplicated().sum()

71

In [253]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [254]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass
    

In [255]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [256]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [257]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [258]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [259]:
def calculate(data, column):
    debt_mean = data.groupby(column)['debt'].mean()
    value = data[column].value_counts()
    def percent(x):
        return "{0:.2%}".format(x)
    debt_mean = debt_mean.apply(percent)
    result = pd.concat([debt_mean, value], axis=1)
    result.columns = ['% задолженности по кредиту', 'количество записей']
    return result

result = calculate(data, 'children')
result.index.name = 'Количество детей'

print(result)

# Ваш код будет здесь. Вы можете создавать новые ячейки.

                 % задолженности по кредиту  количество записей
Количество детей                                               
0                                     7.54%               14091
1                                     9.23%                4808
2                                     9.45%                2052
3                                     8.18%                 330
4                                     9.76%                  41
5                                     0.00%                   9


**Вывод:** Наличие детей может влиять на вероятность задолженности по кредиту. 
Заемщики без детей имеют задолженность реже всех 7.54%, это может быть связано с меньшей финансовой нагрузкой на данную группу.
Заемщики с тремя детьми имеют задолженность в 8.18% случаев, что может быть связано с высокой осознанностью в обращении с финансами, и возможными льготами для многодетных семей.
Заемщики с одним или двумя детьми имеют задолженность в 9.23% и 9.45% случаев, что на 22% и на 25% соответственно выше чем у заемщиков без детей, возможно на это влияет высокая финансовая нагрузка и снижена осознанность в обращении с финансами.
Обратим внимание, данные по семьям с четырьмя и пятью детьми недостаточно репрезентативные для окончательных выводов из-за малого размера выборки. 
Общий вывод: с увеличением количества детей в семье процент задолженности по кредитам в целом имеет тенденцию к росту, это мы видим на примере заемщиков без детей, с одним и двумя детьми, что может быть связано с увеличением финансовой нагрузки. Однако семьи с тремя детьми имеют задолженность в 8.18%, возможно это связано с высокой осознанностью в обращении с финансами, и возможными льготами для многодетных семей.



#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [310]:
family_result = calculate(data, 'family_status')
family_result.index.name = 'Семейное положение'

print(family_result)


# Ваш код будет здесь. Вы можете создавать новые ячейки.

                      % задолженности по кредиту  количество записей
Семейное положение                                                  
Не женат / не замужем                      9.76%                2796
в разводе                                  7.06%                1189
вдовец / вдова                             6.62%                 951
гражданский брак                           9.31%                4134
женат / замужем                            7.56%               12261


**Вывод:** Семейное положение явно влияет на возврат кредита в срок. Одинокие люди и люди в браке вынуждены больше уделять внимания  к финансам и кредитным обязательствам
Самый низкий процент задолженности 6.62% у группы "вдовец / вдова", за ними идут две группы "в разводе" и "женат / замужем" с  7.06% и 7.56% соответственно.  
С большим отрывом от них в негативном смысле идут группы с наибольшим процентом имевших задолженности по кредиту, группы "гражданский брак" и "Не женат / не замужем", с 9.31% и 9.76% соответственно.

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [318]:
def categorize_income2(income):
    if 0 <= income <= 30000:
        return 'от 0 до 30000'
    elif 30001 <= income <= 50000:
        return 'от 30001 до 50000'
    elif 50001 <= income <= 100000:
        return 'от 50001 до 100000'
    elif 100001 <= income <= 150000:
        return 'от 100001 до 150000'
    elif 150001 <= income <= 200000:
        return 'от 150001 до 200000'
    elif 200001 <= income <= 400000:
        return 'от 200001 до 400000'
    elif 400001 <= income <= 1000000:
        return 'от 400001 до 1000000'
    elif income >= 1000001:
        return 'выше 1000000'

data['new_total_income_category'] = data['total_income'].apply(categorize_income2)


new_total_income_category_ordered = calculate(data,'new_total_income_category')
new_total_income_category_ordered = new_total_income_category_ordered.rename_axis('Доход')
desired_order = ['от 0 до 30000','от 30001 до 50000','от 50001 до 100000','от 100001 до 150000',
                    'от 150001 до 200000','от 200001 до 400000','от 400001 до 1000000','выше 1000000']

new_total_income_category_ordered = new_total_income_category_ordered.reindex(desired_order)

print(new_total_income_category_ordered)

                     % задолженности по кредиту  количество записей
Доход                                                              
от 0 до 30000                             9.09%                  22
от 30001 до 50000                         6.02%                 349
от 50001 до 100000                        8.13%                4073
от 100001 до 150000                       8.71%                7110
от 150001 до 200000                       8.51%                4738
от 200001 до 400000                       7.20%                4511
от 400001 до 1000000                      5.77%                 503
выше 1000000                              8.00%                  25


**Вывод:** Для увелечения наглядности выполнена разделение выборок на новые категории по уровню дохода. Так же обратим внимание, категория с доходом от 0 до 30000 и выше 1000000 недостаточно репрезентативные для окончательных выводов из-за малого размера выборки. 
Общий вывод: уровень дохода влияет на  возврат кредита в срок, самый низкий процент задолженности, наблюдается у группы с доходом от 30001 до 50000 и у группы с доходом от 400001 до 1000000.
Лучшие показатели у категории от 400001 до 1000000 - 5.77%, так же хороший показатель у категории от 200001 до 400000 - 7.20%, высокие доходы помогают отдать кредит в срок, на удивление категория с доходом от 30001 до 50000 имеет за должность по кредитам в 6.02% случая, что является тоже хорошим показателем. Самая многочисленная группа заемщиков с доходом от 50001 до 200000 имеет задолженность от 8.13% до 8.71% случаев, что можно считать средним уровнем задолженности, эти группы испытывают наибольшею финансовую нагрузку.




#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [1]:
purpose_category_result = calculate(data, 'purpose_category')
purpose_category_result.index.name = 'Цель получения кредита'
print(purpose_category_result)

# Ваш код будет здесь. Вы можете создавать новые ячейки.

NameError: name 'calculate' is not defined

**Вывод:**  В этой части мы тоже видим зависимость.
Самый низкий процент задолженности у группы "операции с недвижимостью" и "проведение свадьбы", что составляет 7.26% и 7.91% соответственно , эти кредиты берутся наиболее обдуманно и взвешенно. Кредиты на покупку недвижимости часто имеют залог в виде самой недвижимости, что снижает риск.
Так же мы видим, что у группы "получение образования" и "операции с автомобилем" довольно высокий процент задолженности по кредиту, 9.25% и 9.35% соответственно.

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ: Пропуски в данных могут возникать по различным причинам, включая отсутствие информации, ошибки ввода данных, технические проблемы, неправильное представление данных, отсутствие ответов и случайные факторы. 

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* Заполнение пропусков медианным значением для количественных переменных рекомендуется из-за его устойчивости к выбросам, способности сохранять распределение данных, незначительного влияния на ключевые статистические показатели и простоты применения, что делает его эффективным и универсальным методом для обработки недостающих данных.

### Шаг 4: общий вывод.

Напишите ваш общий вывод.

На этапе  предобработки, мы заполнили пропуски, исключили и переработали аномалии значения, которые не отражают действительность и появились по какой-то ошибке. Пропуски  с данными о доходах в столбце `total_income` и с днями трудового стажа в столбце  `days_employed`  заполнены медианным значением с группировкой по типу дохда из столбца `income_type`,  как наиболее влияющее на результат. Для большей дитализации, и уменьшения несбалансированности выборок все заемщики  разделены на 8 категорий по уровнем дохода. Кредит разделены на 4 категории, согласно назначению кредита. 
    В процессе исследования мы отвечали на следующие вопросы:
    1. Есть ли зависимость между количеством детей и возвратом кредита в срок?
Зависимость есть, с увеличением количества детей в семье процент задолженности по кредитам в целом имеет тенденцию к росту! Что возможно связано увеличивающийся финансовой нагрузкой. Заемщики без детей имеют задолженность реже всех в 7.54% случаях, с одним ребенком в 9.23% случаях и двумя детьми имеют задолженность в 9.45%, что является самым большим результатом и на 25% больше чем у заемщиков без детей. Однако семьи с тремя детьми имеют задолженность в 8.18%, что возможно связано с субсидиями для многодетных семей и высокой финансовой ответственностью. Обратим внимание, данные по семьям с четырьмя и пятью детьми недостаточно репрезентативные для окончательных выводов из-за малого размера выборки. 
    2. Есть ли зависимость между семейным положением и возвратом кредита в срок?
Семейное положение влияет на возврат кредита в срок. Лучше всех относятся к кредитным обязательствам одинокие люди после брака и лица находящиеся в браке. Самый низкий процент не возврата у группы людей "вдовец / вдова" 6.62%, за ними следуют "в разводе" 7.06%, "женат / замужем" 7.56%. Что возможно связано с грамотным управлением финансами и высокой финансовой ответственностью. С большим отрывом от них в негативном смысле идут группы с наибольшим процентом имевших задолженности по кредиту, группы "гражданский брак" и "Не женат / не замужем", с 9.31% и 9.76% соответственно. Разница между  группой с меньшим процентом задолженности "вдовец / вдова"  и  группами с самым большим долгом "гражданский брак" и "Не женат / не замужем" составляет 40% и 47% соответственно.
    3. Есть ли зависимость между уровнем дохода и возвратом кредита в срок?
Уровень дохода влияет на возврат кредита в срок, самый низкий процент задолженности, наблюдается у группы с доходом  от 400001 до 1000000 5.77% не возврата, с уменьшением дохода есть тенденция к увеличению процента не возврата кредитов. Что связано с увеличением финансовой нагрузки. Есть одно исключение из этой закономерности лица с доходом от 30001 до 50000 с 6.02% не возврата, что возможно связано с высокой финансовой ответственностью и грамотностью. Самая многочисленная группа заемщиков с доходом от 50001 до 200000 имеет задолженность от 8.13% до 8.71% случаев. Разница по проценту не возврата между самой лучшей группой с доходом от 400001 до 1000000  и худшей с доходом от 100001 до 150000 составляет 51%. Так же обратим внимание, категория с доходом от 0 до 30000 и выше 1000000 недостаточно репрезентативные для окончательных выводов из-за малого размера выборки.
    4. Как разные цели кредита влияют на его возврат в срок?
Цель кредита влияет на процент не возврата. Ниже всего процент не возврата группы "операции с недвижимостью"  7.26% и "проведение свадьбы" 7.91%, эти цели являются, как правило, наиболее обдуманные и взвешенными. Не мало важно, что обе этих группы является наименее рискованными так как, кредиты на покупку недвижимости часто имеют залог в виде самой недвижимости и более низкий процент, а сумма кредита на свадьбу самая низкой по сравнению с другими группами. Так же мы видим, довольно высокий процент задолженности по кредиту у групп "получение образования" 9.25% и "операции с автомобилем" 9.35%, это на 25% и на 28% выше самой низкой группы.
    Подведем итог из данных которые мы проанализировали:
Прослеживается закономерность, увеличение финансовой нагрузки на заемщика негативно сказывается на возврате кредита в срок. 
Идеальный "портрет" заемщика - заемщик состоявший в официальном браке или бывший в браке, без детей, который берет кредит с целью покупки квартиры или брака, с доходом от 400001 до 1000000.
Заемщик  с самой низкой финансовой ответственностью - заемщик ни когда не состоявший в браке, с одним или двумя детьми, который берет кредит с целью покупки машины или на образование, с доходом от 50000 до 200000.
    Для дальнейшего исследования необходимо проверить как оставшиеся категории, такие как общий трудовой стаж, возраст, уровень образования, пол клиента, тип занятости, и их комбинации влияют на возврат кредита в срок.